src/graphics/Window.zig

```diff
@@ -349,17 +583,174 @@ pub fn init() void { // MARK: init()
+}
+var controllerMappingsDownloaded: bool = false;
+var downloadControllerMappingsThread: ?std.Thread = null;
+pub fn controllerMappingsDownloading() bool {
+	return !controllerMappingsDownloaded and downloadControllerMappingsThread != null;
+}
+fn downloadControllerMappingsThreadFunc(curTimestamp: i128) void {
+	var client: std.http.Client = .{.allocator = main.globalAllocator.allocator};
+	var list = std.ArrayList(u8).init(main.globalAllocator.allocator);
+	if (client.fetch(.{
+		.method = .GET,
+		.location = .{.url = "https://raw.githubusercontent.com/mdqinc/SDL_GameControllerDB/master/gamecontrollerdb.txt"},
+		.response_storage = .{ .dynamic = &list }
+	}) catch null) |_| {
+		if (files.openDir(".") catch null) |dir| {
+			dir.write("gamecontrollerdb.txt", list.items) catch unreachable;
+			const timeStampStr = std.fmt.allocPrint(main.globalAllocator.allocator, "{d}", .{curTimestamp}) catch unreachable;
+			dir.write("gamecontrollerdb.stamp", timeStampStr) catch unreachable;
+			main.globalAllocator.free(timeStampStr);
+		}
+	}
+	list.deinit();
+	client.deinit();
+	controllerMappingsDownloaded = true;
+
+}
+pub fn downloadControllerMappings() void {
+
+	var needDownload: bool = false;
+	const curTimestamp = std.time.nanoTimestamp();
+	if (settings.downloadControllerMappings) {
+		var timestamp: i128 = 0;
+		if (files.openDir(".") catch null) |dir| {
+			var _dir: files.Dir =dir;
+			if (_dir.hasFile("gamecontrollerdb.stamp")) {
+				const stamp = _dir.read(main.globalAllocator, "gamecontrollerdb.stamp") catch unreachable;
+				if (std.fmt.parseInt(i128, stamp, 16) catch null) |newTimestamp| {
+					timestamp = newTimestamp;
+				}
+				main.globalAllocator.free(stamp);
+			}
+		}
+		const delta = curTimestamp - timestamp;
+		const deltaDays = @divTrunc(delta, std.time.ns_per_day);
+		needDownload = deltaDays >= 7;
+	}
+
+	if (settings.downloadControllerMappingsWhenUnrecognized) {
+		for (0..c.GLFW_JOYSTICK_LAST) |jsid| {
+			if ((c.glfwJoystickPresent(@intCast(jsid)) != 0) and (c.glfwJoystickIsGamepad(@intCast(jsid)) == 0)) {
+				needDownload = true;
+				break;
+			}
+		}
+	}
+	controllerMappingsDownloaded = !needDownload;
+	if (needDownload) {
+		downloadControllerMappingsThread = std.Thread.spawn(.{}, downloadControllerMappingsThreadFunc, .{curTimestamp}) catch null;
+		if (downloadControllerMappingsThread == null) {
+			std.log.err("Couldn't spawn controller mapping download thread!", .{});
+			controllerMappingsDownloaded = true;
+		}
+	}
+}
+pub fn updateControllerMappings() void {
+	if (std.fs.selfExeDirPathAlloc(main.globalAllocator.allocator) catch null) |selfExeDirPath| {
+		if (std.fs.path.join(main.globalAllocator.allocator, &.{selfExeDirPath, "gamecontrollerdb.txt"}) catch null) |mappings_path| {
+			if (std.fs.openFileAbsolute(mappings_path, .{.mode = .read_only}) catch null) |file| {
+				const data = main.globalAllocator.alloc(u8, file.getEndPos() catch 0);
+				const size = file.readAll(data) catch 0;
```

Comment:
Better use `main.files.read` instead of doing all this boilerplate to read the file.
This is less error prone (like for example here you forgot to close the file).
Also stackAllocator